File: src/java/leetcode/editor/en/P133_CloneGraph.java
package java.leetcode.editor.en;
import java.util.*;
import java.util.function.*;
import java.leetcode.editor.en.definition.*;

public class P133_CloneGraph {
//leetcode submit region begin(Prohibit modification and deletion)
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node node) {
        if (node == null) return null;
        Map<Node, Node> map = new HashMap<>();
        Queue<Node> queue = new ArrayDeque<>();
        queue.offer(node);
        map.put(node, new Node(node.val));
        while (!queue.isEmpty()) {
            Node cur = queue.poll();
            for (Node child : cur.neighbors) {
                if (!map.containsKey(child)) {
                    map.put(child, new Node(child.val));
                    queue.offer(child);
                }
                map.get(cur).neighbors.add(map.get(child));
            }
        }
        return map.get(node);
    }
}
//leetcode submit region end(Prohibit modification and deletion)
}

File: src/kt/leetcode/editor/en/P133_CloneGraph.kt
package kt.leetcode.editor.en;
import java.util.*;
import java.util.function.*;
import kt.leetcode.editor.en.definition.*;

public class P133_CloneGraph {
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var neighbors: ArrayList<Node?> = ArrayList<Node?>()
 * }
 */

class Solution {
    private val map = HashMap<Node, Node>()
    fun cloneGraph(node: Node?): Node? {
        if (node == null) return null
        if (map[node] == null) {
            map[node] = Node(node.`val`)
            for (child in node.neighbors) {
                map[node]!!.neighbors.add(cloneGraph(child)!!)
            }
        }
        return map[node]
    }
}
//leetcode submit region end(Prohibit modification and deletion)
}
